### ISSUE TITLE
space element > 0?

### ISSUE THREAD

**Issue by MatthiasJE** [odlgroup/odl#931]
Should the use of "`x > 0`" be allowed where x is some element of a space X, e.g. `X = odl.uniform_discr(min_pt=[-1, -1], max_pt=[1, 1], shape=[128, 128])`? Most spaces are not ordered so you could argue that it should not be defined. However, it is common to expect as an output an element of `X` where each location checks if `x > 0`, e.g. `X.element(x.asarray() > 0))`. In any case, currently ODL gives out a single `bool`. Is this a bug or do you want this behavior? I find it rather counter intuitive and would prefer the latter behavior (in line with numpy and MATLAB for vectors, matrices, etc)

Edit: @adler-j improved formatting.

**Comment by adler-j** [comment]
### Clarification of the problem

The current behaviour can be seen from this Minimal Working Example:

```python
>>> X = odl.rn(3)
>>> x = X.one()
>>> x < 0
True
>>> x < 'some string'
True
```

This is clearly a bug that should be fixed. Specifically this is related to [a very weird part of the python 2 standard](https://docs.python.org/2/reference/datamodel.html#object.__cmp__) that specifies 

> If no `__cmp__()`, `__eq__()` or `__ne__()` operation is defined, class instances are compared by object identity (“address”)

It seems this is "fixed" in python 3. I.e. while Python 2 gives

```python
>>> 2 < '1'
True
```

Python 3 gives the much more sensible answer:

```python
>>> 2 < '1'
Traceback (most recent call last):
  File "python", line 1, in <module>
TypeError: unorderable types: int() < str()
```

### Regarding `x < 1`

After the above bug has been fixed, the question of how to handle expressions like `x < 1` still remain. As @MatthiasJE notes, both matlab and numpy handle this by returning the pointwise answer for all comparsions.

We have an interesting question of consistency in that one would expect `x < y` to be similar to `x == y`, but for equality, ODL actually has a well defined value and it is a single bool, in contrast to numpy. At the time we added this, it was selected given that ODL should emphasize that vectors are objects "in their own right" and not just tuples of objects.

So we do, to me, basically have three options

* Disable `x < 1`, let users use `np.less(x, 1)` instead if they want pointwise comparsions.
* Implement `x < 1` as pointwise comparsion, i.e. `np.less(x, 1)`.
* Implement `x < 1` as global comparsion, i.e.`np.all(np.less(x, 1))`.

I'll have to think of which one is the best, and any input is ofc nice.

**Comment by kohr-h** [comment]
Nice summary.

Even shorter summary:

1. Stop using Python 2.
2. Do we want to re-implement Numpy? I'd say no.
3. What makes things more complicated than necessary is that `x[bool_array]` doesn't work in ODL, and we probably support will depend on the backend in the future. We can think about wrapping this.
   In the meanwhile, the workaround is
   ```python
   from odl.util import writable_array
   with writable_array(x) as x_arr:
      x_arr[bool_array] = some_other_array
   ```
   To get the boolean array, you can use Numpy functions as @adler-j wrote, e.g. `np.greater(x, 0)`, but the result will be an ODL object, so you need to append `.asarray()`, too.

Having written the above, I realize it's nontrivial and it would be good to support indexing with boolean ODL tensors. Something like this (looking a bit into the future):
```python
def __getitem__(self, indcs):
    if isinstance(indcs, Tensor):
        array = self.asarray()[indcs.asarray()]
        space = # appropriate tensor space from array
        return space.element(array)
    else:
        # current implementation

def __setitem__(self, indcs, values):
    if isinstance(indcs, Tensor):
        with writable_array(self) as self_arr:
        self_arr[indcs.asarray()] = np.asarray(values)
    else:
        # current implementation
```

**Comment by kohr-h** [comment]
And as an addition, of the three options mentioned by @adler-j, I prefer the last option, simply because it's the only one that is compatible with the `==` operation. It's probably also what mathematicians would immediately think when they see `x > 0` where `x` is a vector.

How do you see this @MatthiasJE?

The other option (elementwise) can be implemented anyway using Numpy, so we don't really opt totally against it.

**Comment by MatthiasJE** [comment]
Well, the question is so simple but to me there is no clear answer. Mathematically, I think it depends on the context what to expect. If f is a vector, then I would either expect that f > 0 is a boolean or a bool vector. However, if f is a function, then I would expect from f > 0 rather a set of all x in the domain of f such that f(x) > 0. This is what find(f > 0) in MATLAB does. Of course this can be represented be a vector again and we are back to f > 0 being a bool vector. In any case, to me there does not seem to be a clear answer to this question.

Next to mathematical notation, it might be convenient if a notation has a certain meaning. As numpy and MATLAB  have the same convention, most people (including mathematicians) would expect a similar behavior of ODL I suppose. 

Regarding your comments @kohr-h:
1. Stop using Python 2: It is not that easy. There are libraries that do not support (yet) python 3 so if you need them, you are stuck.
2. Do we want to re-implement Numpy? I'd say no: I would not re-implement numpy but I strongly prefer if ODL is compatible with numpy in the sense that many algorithms can be used with either numpy or ODL. Of course this is not an issue here as one can use np.greater(x,0) for both ODL and numpy. However, it would be more convenient if also other numpy notations can be used. The big question is, do you want ODL to replace other code or should it work alongside with other code? I would vote for the latter.

Regarding the three options of @adler-j:
I wanted to use x > 0 to be used as option 2 because I am used to it from MATLAB. With all being said now x > 0 is ambiguous and it is always dangerous to use a notation that can be misunderstood so I would vote against option 2 and 3 which leaves us with option 1?

**Comment by adler-j** [comment]
> Stop using Python 2: It is not that easy. There are libraries that do not support (yet) python 3 so if you need them, you are stuck.

Sadly this is the case for many of us, and this is partly why I'm still using python 2 on my windows 7 machine myself.

A nice alternative is however to use conda environments, then you can select what python version to use for what application. If your python 2 dependency is not used when you do inverse problems with ODL, you could create a python 3 environment and use that. That would also enable some other nice python 3 stuff (e.g. some optimizations and some nice syntax). I do this myself on occasion, e.g. just yesterday i created a python 3.5 environment in order to install tensorflow.

> Do we want to re-implement Numpy? I'd say no: I would not re-implement numpy but I strongly prefer if ODL is compatible with numpy in the sense that many algorithms can be used with either numpy or ODL. Of course this is not an issue here as one can use np.greater(x,0) for both ODL and numpy. However, it would be more convenient if also other numpy notations can be used. The big question is, do you want ODL to replace other code or should it work alongside with other code? I would vote for the latter.

Here I agree with your conclusions, and we have actually put quite a bit of work towards this in ODL. The goal so far has been:

* Support numpy style functionality (where practical)
* Do not re-implement all of numpy.

The reasoning has been, as you note, that numpy is **the** standard for python, and as a library we need to interact nicely with it. On the other hand, we don't have infinite development resources and thus we cannot rewrite all of numpy. I personally feel the current solution w.r.t numpy ufuncs and arrays in general is quite successful.

> The big question is, do you want ODL to replace other code or should it work alongside with other code? I would vote for the latter.

Agree, and this has actually been one of the drivers in ODL design. In ODL (in contrast to many other libraries) we do not want to do everything for users, but we want to be used where appropriate. In the (upcoming) ODL article we even have a whole subsection called "Avoid user lock-in".

> I wanted to use x > 0 to be used as option 2 because I am used to it from MATLAB. With all being said now x > 0 is ambiguous and it is always dangerous to use a notation that can be misunderstood so I would vote against option 2 and 3 which leaves us with option 1?

I agree with your sentiment here, and it seems @kohr-h does too. In that case, I feel we can close this issue now that your code has been fixed by #933:

```python
>>> X = odl.rn(3)
>>> x = X.one()
>>> x < 0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "e:\github\odl\odl\set\space.py", line 810, in __cmp__
    ''.format(self.__class__.__name__, type(other)))
TypeError: unorderable types: NumpyFnVector, <type 'int'>
```

**Comment by kohr-h** [comment]
> Stop using Python 2: It is not that easy. There are libraries that do not support (yet) python 3 so if you need them, you are stuck.

Sad truth. My point was more that Python 2 has some weird quirks that were debatable design choices, and whenever possible, one should stay away from it. Tough fate if you happen to need packages that haven't been ported.



> Do we want to re-implement Numpy? I'd say no: I would not re-implement numpy but I strongly prefer if ODL is compatible with numpy in the sense that many algorithms can be used with either numpy or ODL. Of course this is not an issue here as one can use np.greater(x,0) for both ODL and numpy. However, it would be more convenient if also other numpy notations can be used. The big question is, do you want ODL to replace other code or should it work alongside with other code? I would vote for the latter.

Agreed. In some sense, though, the `x == y` implementation in ODL already points into a certain direction, so here it would actually inconsistent to handle different comparison operators differently.
That said, I don't think comparison of ODL vectors is a heavily used feature, and reconsidering that decision is not out of question.
Then again, comparing vectors entry by entry is the "right" way of comparing in finite-dimensional vector spaces (apart from `space.dist(x, y) == 0`, which is the default), and that context is an actual difference between ODL and Numpy.

In the long run, the shortest possible syntax will probably be
```python
>>> x = odl.rn(3).element([-1, 0, 1])
>>> x[np.greater(x, 0)] = 0
>>> x
rn(3).elment([-1.0, 0.0, 0.0])
```
which would already be better than now.

So for my part, since we all agree that there's no ideal solution and that the concrete issue at hand has been fixed, this can be closed.
